File: week4&5/Ariling/boj-7579-앱/main.py
import sys
input = sys.stdin.readline

[N, M] = list(map(int, input().split()))
byteArr = list(map(int, input().split()))
costArr = list(map(int, input().split()))
byteArr.insert(0, 0)
costArr.insert(0,0)
totalCost = sum(costArr)
result = sum(costArr)
dp = [[0] * (totalCost + 1) for _ in range(N+1)]
for i in range(1, N+1):
    byte = byteArr[i]
    cost = costArr[i]
    for j in range(totalCost + 1):
        if j < cost:
            dp[i][j] = dp[i-1][j]
        else:
            dp[i][j] = max(byte+dp[i-1][j-cost], dp[i-1][j])
        if dp[i][j] >= M:
            result = min(result, j)
if M == 0:
    print(0)
else:
    print(result)

File: week4&5/Ariling/boj-7579-앱/main.js
const fs = require("fs");
const [firstLine, ...input] = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");
// 처음에 이분탐색으로 했다가 최적해 때문에 망... 이건 냅색이라고 한다..
const [count, needMemory] = firstLine.split(" ").map(Number);
const byteArr = input[0].split(" ").map(Number);
const rateArr = input[1].split(" ").map(Number);
byteArr.unshift(0);
rateArr.unshift(0);
function solution() {
  const rateSum = rateArr.reduce((a, b) => a + b, 0);
  let result = rateArr.reduce((a, b) => a + b, 0);
  const dp = Array(count + 1)
    .fill()
    .map(() => Array(rateSum + 1).fill(0));
  for (let i = 1; i <= count; i++) {
    const byte = byteArr[i];
    const rate = rateArr[i];
    for (let j = 0; j <= rateSum; j++) {
      if (j < rate) {
        dp[i][j] = dp[i - 1][j];
      } else {
        dp[i][j] = Math.max(byte + dp[i - 1][j - rate], dp[i - 1][j]);
      }
      if (dp[i][j] >= needMemory) {
        result = Math.min(result, j);
      }
    }
  }
  needMemory !== 0 ? console.log(result) : console.log(0);
}
solution();
